upgrade build from sbt 0.13 -> sbt 1 [ci: last-only]

SethTisue opened the pull request:
~~wippity wip wip. let's get the ball rolling, at least~~

UPDATE: this is nearing completion.

SethTisue commented:
@dwijnand @eed3si9n `Load` is now in `sbt.internal` and is inaccessible, so the `setup` method in `project/ScriptCommands.scala` fails to compile with "not found: value Load". can you advise on how this method should now be defined...?

dwijnand commented:
There is no migration path, outside of requesting an API be made public.

In this case I think what we should do is fix `Extracted#append`, or add another method. We can try and do that for sbt 1.1.1, both Martynas and Justin (of Jetbrains) have run into the same problem in the last 10 days.

dwijnand commented:
https://github.com/sbt/sbt/pull/3865

SethTisue commented:
sbt 1.1.1 is out and includes sbt/sbt#3865

dwijnand commented:
eh. I meant 1.1.1. we'll fix it when we squash later

SethTisue commented:
/rebuild

SethTisue commented:
thanks Dale for taking this over! I've squashed and made you author of the resulting commit

SethTisue reviewed (commented):
LGTM, except possibly the removal of `antStyle` — team (@retronym?), I think I might recall that actually still being used sometimes...? if so, maybe we could give it a better name.

also Dale, you commented out `-Xlint` in `project/plugins.sbt`, would leaving it enabled present some special difficulty...?

dwijnand commented:
> LGTM, except possibly the removal of antStyle — team (@retronym?), I think I might recall that actually still being used sometimes...? if so, maybe we could give it a better name.

antStyle was torn out of the incremental compiler, I'm not sure when (or what commit).

> also Dale, you commented out `-Xlint` in `project/plugins.sbt`, would leaving it enabled present some special difficulty...?

it could stay but it would need to be tweaked, for instance make it not tell you about every import that sbt injects into build.sbt but that build doesn't use. stuff like that. I took the easy way out, but I can be more diligent if you'd like.

SethTisue commented:
`-Xlint:-unused,_` makes it manageable.

retronym commented on project/Quiet.scala:
```diff
@@ -8,26 +8,6 @@ object Quiet {
   //
   //   https://github.com/scala/scala-dev/issues/100
   def silenceScalaBinaryVersionWarning = ivyConfiguration := {
-    ivyConfiguration.value match {
```
The warnings we're trying to suppress are still emitted by SBT 1.1, so we'll have to figure out how to port this, or otherwise have a plan to make the warning suppressable in SBT itself.

```
sbt:root> test:compile
[warn] Binary version (2.13.0-pre-SNAPSHOT) for dependency org.scala-lang#scala-library;2.13.0-pre-SNAPSHOT
[warn] 	in org.scala-lang#scala-reflect;2.13.0-pre-SNAPSHOT differs from Scala binary version in project (2.13.0-M3).
[warn] Binary version (2.13.0-pre-SNAPSHOT) for dependency org.scala-lang#scala-library;2.13.0-pre-SNAPSHOT
[warn] 	in org.scala-lang#scala-compiler;2.13.0-pre-SNAPSHOT differs from Scala binary version in project (2.13.0-M3).
[warn] Binary version (2.13.0-pre-SNAPSHOT) for dependency org.scala-lang#scala-reflect;2.13.0-pre-SNAPSHOT
[warn] 	in org.scala-lang#scala-compiler;2.13.0-pre-SNAPSHOT differs from Scala binary version in project (2.13.0-M3).
[warn] Binary version (2.13.0-pre-SNAPSHOT) for dependency org.scala-lang#scala-compiler;2.13.0-pre-SNAPSHOT
[warn] 	in org.scala-lang#scalap;2.13.0-pre-SNAPSHOT differs from Scala binary version in project (2.13.0-M3).
[warn] Binary version (2.13.0-pre-SNAPSHOT) for dependency org.scala-lang#scala-compiler;2.13.0-pre-SNAPSHOT
[warn] 	in org.scala-lang#scala-compiler-interactive;2.13.0-pre-SNAPSHOT differs from Scala binary version in project (2.13.0-M3).
[warn] Binary version (2.13.0-pre-SNAPSHOT) for dependency org.scala-lang#scala-compiler;2.13.0-pre-SNAPSHOT
[warn] 	in org.scala-lang#repl;2.13.0-pre-SNAPSHOT differs from Scala binary version in project (2.13.0-M3).
[warn] Binary version (2.13.0-pre-SNAPSHOT) for dependency org.scala-lang#scala-compiler;2.13.0-pre-SNAPSHOT
[warn] 	in org.scala-lang#scala-compiler-doc;2.13.0-pre-SNAPSHOT differs from Scala binary version in project (2.13.0-M3).
[info] Running scala.tools.docutil.ManMaker fsc, scala, scalac, scaladoc, scalap /Users/jz/code/scala-sbt-1/target/scala-dist/resource_managed/main/doc/tools /Users/jz/code/scala-sbt-1/target/scala-dist/resource_managed/main/genman
[info] Compiling 526 Scala sources and 120 Java sources to /Users/jz/code/scala-sbt-1/build/quick/classes/library ...
```

retronym commented:
We can act on a new incremental compiler warning by deleting the file that contains only comments:

```
[warn] /Users/jz/code/scala-sbt-1/src/reflect/scala/reflect/io/IOStats.scala:1:1:
[warn] Found names but no class, trait or object is defined in the compilation unit.
[warn] The incremental compiler cannot record the dependency information in such case.
[warn] Some errors like unused import referring to a non-existent class might not be reported.
[warn]
[warn] package scala
[warn] ^
```

SethTisue commented:
I took care of removing `IOStats.scala`. @dwijnand can you help with the suppress-binary-warnings stuff? I remember the noise level on that being really high before Jason managed to suppress them

retronym commented:
The performance of a clean build (in a well-warmed up JVM) has gone from 84s up to 101s.

To measure, I use a `cleanClasses` task, defined in `~/.sbt/{1.0,0.13}/plugins/Clean.scala`:

```scala
package io.github.retronym

import sbt._
import Keys._

object CleanClasses extends AutoPlugin {
  override def trigger = allRequirements
  override def requires = sbt.plugins.JvmPlugin
  val cleanClasses = taskKey[Unit]("clean the classes directory")
  
  override lazy val projectSettings = List(Compile, Test).flatMap(c => inConfig(c)(Seq(
    cleanClasses := IO.delete(classDirectory.value)
  )))

}

```

Then run the following repeatedly in both the 2.13.x branch (using SBT 0.13) and this PR:

```
sbt:root> ;cleanClasses;test:cleanClasses
...
sbt:root> test:compile
...
[success] Total time: 101 s, completed 16/02/2018 2:33:41 PM
```

retronym commented:
I repeated the compilation/clean cycle a few times, the attached FlightRecorder to collect a profile.

During that time, this error was issued:

```
[info] Compiling 14 Scala sources and 1 Java source to /Users/jz/code/scala-sbt-1/build/quick/classes/partest-extras ...
[info] Done compiling.
[error] ## Exception when compiling 0 sources to /Users/jz/code/scala-sbt-1/target/partest-extras/test-classes
[error] Could not create directory /Users/jz/code/scala-sbt-1/target/scala-2.13.0-M3/classes.bak
[error] scala.sys.package$.error(package.scala:27)
[error] sbt.io.IO$.createDirectory(IO.scala:227)
[error] sbt.internal.inc.ClassFileManager$TransactionalClassFileManager.<init>(ClassFileManager.scala:73)
[error] sbt.internal.inc.ClassFileManager$.transactional(ClassFileManager.scala:67)
[error] sbt.internal.inc.ClassFileManager$.getDefaultClassFileManager(ClassFileManager.scala:32)
[error] sbt.internal.inc.ClassFileManager$.getClassFileManager(ClassFileManager.scala:39)
[error] sbt.internal.inc.Incremental$.manageClassfiles(Incremental.scala:128)
[error] sbt.internal.inc.Incremental$.compile(Incremental.scala:75)
[error] sbt.internal.inc.IncrementalCompile$.apply(Compile.scala:61)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:309)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:267)
[error] sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:158)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:237)
[error] sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:68)
[error] sbt.Defaults$.compileIncrementalTaskImpl(Defaults.scala:1429)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$1(Defaults.scala:1403)
[error] scala.Function1.$anonfun$compose$1(Function1.scala:44)
[error] sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:39)
[error] sbt.std.Transform$$anon$4.work(System.scala:66)
[error] sbt.Execute.$anonfun$submit$2(Execute.scala:262)
[error] sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:16)
[error] sbt.Execute.work(Execute.scala:271)
[error] sbt.Execute.$anonfun$submit$1(Execute.scala:262)
[error] sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:174)
[error] sbt.CompletionService$$anon$2.call(CompletionService.scala:36)
[error] java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[error] java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[error] java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[error] java.lang.Thread.run(Thread.java:748)
[error]
[info] Compiling 161 Scala sources and 2 Java sources to /Users/jz/code/scala-sbt-1/target/junit/test-classes ...
```

smarter commented:
Back when it was first introduced, I noted that class-based dependency tracking introduced a big overhead for scalac : https://github.com/sbt/sbt/issues/1104#issuecomment-199360458, this is probably still the case. At the time, this overhead didn't seem to affect dotty, I can't test dotty with sbt 1 right now but we'll probably be done with porting dotty to sbt 1 sometimes in the next century or two (current progress at https://github.com/lampepfl/dotty/pull/3872, but this has been ongoing for over a year). once this is done I'll try to run some comparison benchmarks too.

smarter commented:
> Oh, I'll bet the problem is the fact we're setting baseDirectory in the submodules to the root project's baseDirectory. SBT creates the incremental compiler's cache relative to that directory. We probably instead should customize ForkOptions.workingDirectory.

In dotty we do:
```scala
baseDirectory in (Compile, run) := baseDirectory.value / "..",
baseDirectory in Test := baseDirectory.value / ".."
```
and haven't run into problem yet (and I don't see anything suspicious in the root target directory)

retronym commented:
>  @dwijnand can you help with the suppress-binary-warnings stuff? I remember the noise level on that being really high before Jason managed to suppress them

Looks like we can suppress the check with `scalaModuleInfo := scalaModuleInfo.value.withCheckExplicit(false)` configured at the right places.

retronym commented:
`consoleProject` appears to have an incompatibility with Scala 2.13. Not a deal breaker, but we need to get it fixed as it will end up affecting end users.

```
sbt:root> consoleProject
[info] Non-compiled module 'compiler-bridge_2.13.0-M2' for Scala 2.12.4. Compiling...
/var/folders/tz/p8vd07wn7wxck3b9v54grlzw0000gp/T/sbt_d61bd5d9/xsbt/Compat.scala:13: error: object shell is not a member of package scala.tools.nsc.interpreter
import scala.tools.nsc.interpreter.shell.ReplReporterImpl
                                   ^
/var/folders/tz/p8vd07wn7wxck3b9v54grlzw0000gp/T/sbt_d61bd5d9/xsbt/Compat.scala:22: error: not found: type ReplReporterImpl
    new ReplReporterImpl(settings, writer)
        ^
/var/folders/tz/p8vd07wn7wxck3b9v54grlzw0000gp/T/sbt_d61bd5d9/xsbt/ConsoleInterface.scala:12: error: object shell is not a member of package scala.tools.nsc.interpreter
import scala.tools.nsc.interpreter.shell.{ ILoop, ShellConfig, ReplReporterImpl }
                                   ^
/var/folders/tz/p8vd07wn7wxck3b9v54grlzw0000gp/T/sbt_d61bd5d9/xsbt/ConsoleInterface.scala:42: error: not found: type ILoop
    val loop = new ILoop(ShellConfig(interpreterSettings)) {
                   ^
/var/folders/tz/p8vd07wn7wxck3b9v54grlzw0000gp/T/sbt_d61bd5d9/xsbt/ConsoleInterface.scala:45: error: not found: type ReplReporterImpl
          val reporter = new ReplReporterImpl(interpreterSettings)
                             ^
/var/folders/tz/p8vd07wn7wxck3b9v54grlzw0000gp/T/sbt_d61bd5d9/xsbt/ConsoleInterface.scala:46: error: not found: value intp
          intp = new IMain(interpreterSettings, reporter) {
          ^
/var/folders/tz/p8vd07wn7wxck3b9v54grlzw0000gp/T/sbt_d61bd5d9/xsbt/ConsoleInterface.scala:51: error: not found: value intp
          intp.setContextClassLoader()
          ^
/var/folders/tz/p8vd07wn7wxck3b9v54grlzw0000gp/T/sbt_d61bd5d9/xsbt/ConsoleInterface.scala:53: error: value createInterpreter is not a member of AnyRef
          super.createInterpreter(interpreterSettings)
                ^
/var/folders/tz/p8vd07wn7wxck3b9v54grlzw0000gp/T/sbt_d61bd5d9/xsbt/ConsoleInterface.scala:56: error: not found: value intp
          intp.beQuietDuring(intp.bind(id, value.asInstanceOf[AnyRef].getClass.getName, value))
          ^
/var/folders/tz/p8vd07wn7wxck3b9v54grlzw0000gp/T/sbt_d61bd5d9/xsbt/ConsoleInterface.scala:56: error: not found: value intp
          intp.beQuietDuring(intp.bind(id, value.asInstanceOf[AnyRef].getClass.getName, value))
                             ^
/var/folders/tz/p8vd07wn7wxck3b9v54grlzw0000gp/T/sbt_d61bd5d9/xsbt/ConsoleInterface.scala:59: error: not found: value intp
          intp.interpret(initialCommands)
          ^
/var/folders/tz/p8vd07wn7wxck3b9v54grlzw0000gp/T/sbt_d61bd5d9/xsbt/ConsoleInterface.scala:66: error: not found: value intp
          intp.interpret(cleanupCommands)
          ^
12 errors found
[info] Attempting to fetch org.scala-sbt:compiler-bridge_2.13.0-M2:1.1.1.
[error] (run-main-2) Error compiling the sbt component 'compiler-bridge_2.13.0-M2'
[error] Error compiling the sbt component 'compiler-bridge_2.13.0-M2'
[error] 	at sbt.internal.inc.AnalyzingCompiler$.handleCompilationError$1(AnalyzingCompiler.scala:331)
[error] 	at sbt.internal.inc.AnalyzingCompiler$.$anonfun$compileSources$4(AnalyzingCompiler.scala:346)
[error] 	at sbt.internal.inc.AnalyzingCompiler$.$anonfun$compileSources$4$adapted(AnalyzingCompiler.scala:341)
[error] 	at sbt.io.IO$.withTemporaryDirectory(IO.scala:376)
[error] 	at sbt.io.IO$.withTemporaryDirectory(IO.scala:383)
```

retronym commented:
I profiled the warm, clean compile.

I noticed something unusual with busy-waiting from a LMAX disruptor thread involved in async logging, details in https://github.com/sbt/util/pull/68

The new logging implementation also seems to spend an inordinate amount of time in Scala reflection to evaluate type tags to add to log messages, I've raised https://github.com/sbt/util/issues/151 with an initial diagnosis.

Of time taken to compile, the profile suggests around 40% is spent in Zinc's API phase. 

![image](https://user-images.githubusercontent.com/65551/36296034-e2c03fb2-1336-11e8-9912-010149f21371.png)

retronym commented:
@SethTisue I've pushed commits to fix the problem with intermittent failures related to `classes.bak`, and to suppress the spurious warnings about the Scala version.

dwijnand commented on project/build.properties:
```diff
@@ -1 +1 @@
-sbt.version=1.1.1
+sbt.version=1.1.2-LOCAL31
```
oops

SethTisue replied:
I amended Jason's commit to fix this

SethTisue commented:
I think we can merge this once Jenkins signs off.

retronym commented:
I’d personally like to put it on hold until we finish getting out Jenkins
release jobs up and running

SethTisue commented:
rebased. assuming the 2.12.5 release goes well on the new infrastructure, we could revisit this once the next 2.12.x -> 2.13.x merge has happened

SethTisue commented:
looks like this stuff needs forward-porting:

```
[error] /home/jenkins/workspace/scala-2.13.x-validate-main/project/PartestTestListener.scala:24:51: type Value is not a member of object sbt.protocol.testing.TestResult
[error]   override def doComplete(finalResult: TestResult.Value): Unit = ()
[error]                                                   ^
[error] /home/jenkins/workspace/scala-2.13.x-validate-main/project/PartestTestListener.scala:26:58: type Value is not a member of object sbt.protocol.testing.TestResult
[error]   override def endGroup(name: String, result: TestResult.Value): Unit = ()
[error]                                                          ^
[error] two errors found
```

SethTisue commented:
rebased again. @dwijnand can you advise us what the current recommended why to make `clean` clean additional files is...? (we can't override `cleanFilesTask` anymore, it looks like.)

and also on how to handle this warning?

```
/Users/tisue/scala.213/build.sbt:129: warning: constructor ScalaInstance in class ScalaInstance is deprecated (since 1.1.2): Use constructor with loaderLibraryOnly
      val s2 = new ScalaInstance(s.version, s.loader, s.libraryJar, s.compilerJar, s.allJars, Some(s.actualVersion))
               ^
```

not sure what I should be passing for `loaderLibraryOnly`

dwijnand commented:
> we can't override `cleanFilesTask` anymore, it looks like

use `cleanFiles`

for context: `cleanFiles` was a `SettingKey` in sbt 0.13, we made it a `TaskKey` in sbt 1 to fix a bug. then via a Lightbend subscription we back-ported it to sbt 0.13, but called it `cleanFilesTask` to maintain compatibility within the 0.13 series.

> not sure what I should be passing for `loaderLibraryOnly`

I didn't follow the changes intimately but see https://github.com/sbt/zinc/commit/9397b6aaf94ac3cfab386e3abd11c0ef9c2ceaff#diff-b0624a5545fa4f50b411a99e2f5c1d87L143.

something along the lines of

```scala
val loaderLibraryOnly = ClasspathUtilities.toLoader(Vector(s.libraryJar))
```

perhaps

retronym commented:
The performance of clean `compile` and of `reload` both seem okay, about on par with 0.13.17 from a rough test.

> not sure what I should be passing for loaderLibraryOnly

I've pushed a fix for this.

retronym commented:
We've seen this before, somehow the sbt runner script is passing `-foo` option to SBT itself, which thinks its the deprecated `-` command (which installs a user defined error handler).

```
[warn] The `-` command is deprecated in favor of `onFailure` and will be removed in a later version
```

retronym commented:
Next issue:

```
[error] java.io.IOException: PUT operation to URL https://scala-ci.typesafe.com/artifactory/scala-pr-validation-snapshots;build.timestamp=1523331212877/org/scala-lang/scala-library/2.13.0-pre-c342a14-SNAPSHOT/scala-library-2.13.0-pre-c342a14-SNAPSHOT.pom failed with status code 409: Conflict
[error] 	at org.apache.ivy.util.url.AbstractURLHandler.validatePutStatusCode(AbstractURLHandler.java:82)
[error] 	at sbt.internal.librarymanagement.ivyint.GigahorseUrlHandler.upload(GigahorseUrlHandler.scala:191)
[error] 	at org.apache.ivy.util.url.URLHandlerDispatcher.upload(URLHandlerDispatcher.java:82)
[error] 	at org.apache.ivy.util.FileUtil.copy(FileUtil.java:150)
[error] 	at org.apache.ivy.plugins.repository.url.URLRepository.put(URLRepository.java:84)
[error] 	at sbt.internal.librarymanagement.ConvertResolver$LocalIfFileRepo.put(ConvertResolver.scala:365)
[error] 	at org.apache.ivy.plugins.repository.AbstractRepository.put(AbstractRepository.java:130)
[error] 	at sbt.internal.librarymanagement.ConvertResolver$ChecksumFriendlyURLResolver.put(ConvertResolver.scala:117)
[error] 	at sbt.internal.librarymanagement.ConvertResolver$ChecksumFriendlyURLResolver.put$(ConvertResolver.scala:104)
[error] 	at sbt.internal.librarymanagement.ConvertResolver$$anonfun$defaultConvert$lzycompute$1$PluginCapableResolver$1.put(ConvertResolver.scala:164)
[error] 	at org.apache.ivy.plugins.resolver.RepositoryResolver.publish(RepositoryResolver.java:216)
[error] 	at sbt.internal.librarymanagement.IvyActions$.$anonfun$publish$5(IvyActions.scala:499)
[error] 	at sbt.internal.librarymanagement.IvyActions$.$anonfun$publish$5$adapted(IvyActions.scala:498)
[error] 	at scala.collection.TraversableLike$WithFilter.$anonfun$foreach$1(TraversableLike.scala:789)
```

SethTisue commented:
merged latest 2.13.x, but I think this will need another merge or rebase once #6436 lands

SethTisue commented:
@retronym shall we merge? if it seems ready to you, I can clean up the history before merge.

SethTisue commented:
discussed this with @retronym just now:

the argument for caution: "I’d bet we hit a few more problems yet... Big, fat, obvious problems aren’t that bad, as we could revert. But subtler problems could be costly. Maybe we should do some local development of 2.13 PRs atop that change, but rebase them back to 2.13.x before merge. Just to test the waters a bit more... Maybe I’m being overly paranoid"

the argument to expedite this: "it seems to me that now is a good time to pull the trigger, because M4 is looming, but is not yet imminent. the milestones and RCs that follow M4 are only going to be _more_ visible and _more_ widely used than M4, so I think we should bite the bullet now because regressions only become more costly later, and upgrading now increases our chances that we’ll catch them even _before_ M4"

to-do list:
* diff the distros before and after, including JAR contents (using `jardiff` I guess)
* make sure CI and release jobs are all happy
* make sure CI and release jobs still actually fail on all the possible errors
* make sure local development is still productive (inc compiler works)

retronym commented:
I'm seeing a non-null diff in the result of `sbt setupPublishCore clean dist/mkPack`.

For instance:

```
$ jardiff /tmp/lib-reference/scala-library.jar/scala/concurrent/ExecutionContext.class /tmp/lib-new/scala-library.jar/scala/concurrent/ExecutionContext.class
diff --git a/ExecutionContext.class.asm b/ExecutionContext.class.asm
index b501025..b04b5f4 100644
--- a/ExecutionContext.class.asm
+++ b/ExecutionContext.class.asm
@@ -86,17 +86,20 @@
     IFEQ L3
     ALOAD 1
     INVOKESTATIC scala/concurrent/impl/ExecutionContextImpl$.$anonfun$fromExecutor$1 (Lscala/Function1;)Ljava/util/concurrent/ExecutorService;
+    ASTORE 3
     GOTO L4
    L3
     ALOAD 2
     INVOKEVIRTUAL scala/Option.get ()Ljava/lang/Object;
+    ASTORE 3
    L4
+    ALOAD 3
     CHECKCAST java/util/concurrent/Executor
     ALOAD 1
     INVOKESPECIAL scala/concurrent/impl/ExecutionContextImpl.<init> (Ljava/util/concurrent/Executor;Lscala/Function1;)V
     ARETURN
     MAXSTACK = 4
-    MAXLOCALS = 3
+    MAXLOCALS = 4
```

This suggests that the optimizer didn't run in the new build, or maybe ran but didn't work the same.

adriaanm commented:
Rebased.

adriaanm commented:
Hmm, while running under `-Dstarr.version=2.13.0-local-ae43506d`, running into some weird issues. We're on sbt 1.2.3, but the bridge is 1.2.2, and binary version 2.13 (closest published artifact is at https://mvnrepository.com/artifact/org.scala-sbt/compiler-bridge_2.13.0-M2/1.2.2)

```
[info] Attempting to fetch org.scala-sbt:compiler-bridge_2.13:1.2.2.
[warn] 	module not found: org.scala-sbt#compiler-bridge_2.13;1.2.2
[warn] ==== local: tried
[warn]   /Users/adriaan/.ivy2/local/org.scala-sbt/compiler-bridge_2.13/1.2.2/ivys/ivy.xml
[warn] ==== public: tried
[warn]   https://repo1.maven.org/maven2/org/scala-sbt/compiler-bridge_2.13/1.2.2/compiler-bridge_2.13-1.2.2.pom
[warn] ==== local-preloaded-ivy: tried
[warn]   /Users/adriaan/.sbt/preloaded/org.scala-sbt/compiler-bridge_2.13/1.2.2/ivys/ivy.xml
[warn] ==== local-preloaded: tried
[warn]   file:////Users/adriaan/.sbt/preloaded/org/scala-sbt/compiler-bridge_2.13/1.2.2/compiler-bridge_2.13-1.2.2.pom
[warn] 	::::::::::::::::::::::::::::::::::::::::::::::
[warn] 	::          UNRESOLVED DEPENDENCIES         ::
[warn] 	::::::::::::::::::::::::::::::::::::::::::::::
[warn] 	:: org.scala-sbt#compiler-bridge_2.13;1.2.2: not found
[warn] 	::::::::::::::::::::::::::::::::::::::::::::::
[error] ## Exception when compiling 524 sources to /Users/adriaan/git/scala/build/quick/classes/library
[error] The compiler bridge sources org.scala-sbt:compiler-bridge_2.13:1.2.2:compile could not be retrieved.
[error] 
```

eed3si9n commented:
https://github.com/sbt/sbt/blob/v1.2.3/project/Dependencies.scala#L14
sbt 1.2.3 uses Zinc 1.2.2.

retronym commented:
The artifact name `2.13.0-local-ae43506d` is the problem.

From `internal/zinc-ivy-integration/src/main/scala/sbt/internal/inc/ZincComponentCompiler.scala`:

```scala
  private[sbt] def getDefaultBridgeModule(scalaVersion: String): ModuleID = {
    def compilerBridgeId(scalaVersion: String) = {
      scalaVersion match {
        case sc if (sc startsWith "2.10.")       => "compiler-bridge_2.10"
        case sc if (sc startsWith "2.11.")       => "compiler-bridge_2.11"
        case sc if (sc startsWith "2.12.")       => "compiler-bridge_2.12"
        case "2.13.0-M1"                         => "compiler-bridge_2.12"
        case sc if (sc startsWith "2.13.0-pre-") => "compiler-bridge_2.13.0-M2"
        case sc if (sc startsWith "2.13.0-M")    => "compiler-bridge_2.13.0-M2"
        case sc if (sc startsWith "2.13.0-RC")   => "compiler-bridge_2.13.0-M2"
        case _                                   => "compiler-bridge_2.13"
      }
    }
```

`sbt setupPublishCore publishLocal` generate a version number starting with `2.13.0-pre-`, which should use the compiler bridge sources published under 2.13.0-M2.

adriaanm commented:
Thanks, that fixed it. My local.sbt now has:

```
// skip docs for local publishing
publishArtifact in (Compile, packageDoc) in ThisBuild := false

// set version based on current sha, so that you can easily consume this build from another sbt project
baseVersionSuffix := {
  val sha = scala.sys.process.Process("git rev-parse --short HEAD").lines.head
  if (scalaVersion.value.startsWith("2.13")) s"pre-$sha-local"
  else s"bin-$sha-local"
}
```

adriaanm commented:
I'll redo the stability check on the current build and sbt 1.2.3. If I don't get any discrepancies, I assume this PR is ready to go?

adriaanm commented:
Looks like we're good re: stability. Transcript of my due diligence:

```
commit 815d29e (HEAD -> 2.13.x)
Author: Adriaan Moors <adriaan@lightbend.com>
Date:   5 minutes ago

    opt stability test 2

commit eb3e081
Author: Adriaan Moors <adriaan@lightbend.com>
Date:   10 minutes ago

    opt stability test

commit 2e82444 (scala/2.13.x)
```


```
g commit -a --allow-empty -m"opt stability test sbt" ; sbt clean enableOptimizer publishLocal ; g commit -a --allow-empty -m"opt stability test 2" ; sbt clean enableOptimizer publishLocal

cd /tmp/jd

for j in /Users/adriaan/.ivy2/local/org.scala-lang/scala-library/2.13.0-pre-eb3e081-local/jars/scala-library.jar /Users/adriaan/.ivy2/local/org.scala-lang/scala-reflect/2.13.0-pre-eb3e081-local/jars/scala-reflect.jar /Users/adriaan/.ivy2/local/org.scala-lang/scala-compiler/2.13.0-pre-eb3e081-local/jars/scala-compiler.jar; do unzip -o $j ; done

g init .
g add -A .
g commit -m"eb3e081"
```

``` 
for j in /Users/adriaan/.ivy2/local/org.scala-lang/scala-library/2.13.0-pre-815d29e-local/jars/scala-library.jar /Users/adriaan/.ivy2/local/org.scala-lang/scala-reflect/2.13.0-pre-815d29e-local/jars/scala-reflect.jar /Users/adriaan/.ivy2/local/org.scala-lang/scala-compiler/2.13.0-pre-815d29e-local/jars/scala-compiler.jar; do unzip -o $j ; done

g commit -a -m"815d29e"
```


```
g commit -a --allow-empty -m"opt stability test sbt 1" ; sbt clean enableOptimizer publishLocal ; g commit -a --allow-empty -m"opt stability test sbt 1 / 2" ; sbt clean enableOptimizer publishLocal
```

```
commit a223519 (HEAD -> 2.13.x)
Author: Adriaan Moors <adriaan@lightbend.com>
Date:   40 minutes ago

    opt stability test sbt 1 / 2

commit 49bb883
Author: Adriaan Moors <adriaan@lightbend.com>
Date:   49 minutes ago

    opt stability test sbt 1

commit 55ea497
Merge: 815d29e 21439bf
Author: Adriaan Moors <adriaan@lightbend.com>
Date:   49 minutes ago

    Merge remote-tracking branch 'SethTisue/sbt-1' into 2.13.x
```

```
cd /tmp/jd

for j in /Users/adriaan/.ivy2/local/org.scala-lang/scala-library/2.13.0-pre-49bb883-local/jars/scala-library.jar /Users/adriaan/.ivy2/local/org.scala-lang/scala-reflect/2.13.0-pre-49bb883-local/jars/scala-reflect.jar /Users/adriaan/.ivy2/local/org.scala-lang/scala-compiler/2.13.0-pre-49bb883-local/jars/scala-compiler.jar; do unzip -o $j ; done

g commit -a -m"49bb883"

for j in /Users/adriaan/.ivy2/local/org.scala-lang/scala-library/2.13.0-pre-a223519-local/jars/scala-library.jar /Users/adriaan/.ivy2/local/org.scala-lang/scala-reflect/2.13.0-pre-a223519-local/jars/scala-reflect.jar /Users/adriaan/.ivy2/local/org.scala-lang/scala-compiler/2.13.0-pre-a223519-local/jars/scala-compiler.jar; do unzip -o $j ; done

g commit -a -m"a223519"
```

```
g log --stat

commit 2415fb0 (HEAD -> master)
Author: Adriaan Moors <adriaan@lightbend.com>
Date:   11 seconds ago

    a223519:

 META-INF/MANIFEST.MF            | 188 ++++++++++++++++++++--------------------
 compiler.properties             |   8 +-
 interactive.properties          |   8 +-
 library.properties              |   8 +-
 reflect.properties              |   8 +-
 repl-frontend.properties        |   8 +-
 repl.properties                 |   8 +-
 scala-buildcharacter.properties |  10 +--
 scaladoc.properties             |   8 +-
 9 files changed, 127 insertions(+), 127 deletions(-)

commit ef46483
Author: Adriaan Moors <adriaan@lightbend.com>
Date:   62 seconds ago

    49bb883

 META-INF/MANIFEST.MF            | 188 ++++++++++++++++++++--------------------
 compiler.properties             |   8 +-
 interactive.properties          |   8 +-
 library.properties              |   8 +-
 reflect.properties              |   8 +-
 repl-frontend.properties        |   8 +-
 repl.properties                 |   8 +-
 scala-buildcharacter.properties |  10 +--
 scaladoc.properties             |   8 +-
 9 files changed, 127 insertions(+), 127 deletions(-)

commit bb783a5
Author: Adriaan Moors <adriaan@lightbend.com>
Date:   47 minutes ago

    815d29e

 META-INF/MANIFEST.MF            | 188 ++++++++++++++++++++--------------------
 compiler.properties             |   8 +-
 interactive.properties          |   8 +-
 library.properties              |   8 +-
 reflect.properties              |   8 +-
 repl-frontend.properties        |   8 +-
 repl.properties                 |   8 +-
 scala-buildcharacter.properties |  10 +--
 scaladoc.properties             |   8 +-
 9 files changed, 127 insertions(+), 127 deletions(-)

commit 659ff5a
Author: Adriaan Moors <adriaan@lightbend.com>
Date:   54 minutes ago

    eb3e081

 6518 files changed, 5118 insertions(+)

```

adriaanm commented:
Rebased

SethTisue committed:
Upgrade build from sbt 0.13.17 to 1.2.3

SethTisue committed:
Enable more linting in build, fatal warnings too

retronym committed:
Various fixes for sbt 1.2 upgrade

1. Adapt to renaming of ivyScala

2. Be less intrusive with customization of baseDirectory

In our attempt to make forked project take on the working directory
of the root project, rather than of their module, we ended up
causing a clash in the directory used by the incremental compiler
to backup classfiles, which uses a value relative to that base.

The symptom was an intermitten error such as:

```
[error] ## Exception when compiling 0 sources to /Users/jz/code/scala-sbt-1/target/partest-extras/test-classes
[error] Could not create directory /Users/jz/code/scala-sbt-1/target/scala-2.13.0-M3/classes.bak
```

This commit adjusts ForkOptions instead to more directly
achieve our goal.

3. Heed deprecation warning in ScalaInstance tweak

4. Disable finding sources files in project base

Something has changed in SBT 1.x, and our attempt to clear the
sources for scalacheck/compile: weren't successful:

```
sbt:root> show scalacheck / Compile / unmanagedSources
[info] * /Users/jz/code/scala/test/scalacheck/array-new.scala
[info] * /Users/jz/code/scala/test/scalacheck/array-old.scala
[info] * /Users/jz/code/scala/test/scalacheck/CheckCollections.scala
[info] * /Users/jz/code/scala/test/scalacheck/CheckEither.scala
```

What changed? 4502348a12c, earlier in this PR, removed our override
of `baseDirectory`. So sub-projects now have their own sub-directory
as the base (which is the standard SBT behaviour). If these include
source files directly, as is the case with `./test/scalacheck`, these
would be included by virtue of the long standing SBT behaviour to
let you skip the hassle of source directories for toy projects.

This commit disables that behaviour with `sourcesInBase := false`

5. Fixup OSGi test suite after baseDirectory change

6. Let SBT add api.url to the POM

Since https://github.com/sbt/sbt/pull/2262, we don't need to do this.
And attempting to do it results in duplicate properties in the POM file,
which doesn't pass POM validation by artifactory.

7. Use the recommended, and working, `sbt -warn`

In favour of `--warn`, which is broken in SBT 1.x.

Early commands are ones that are run before project load, in this case,
`-warn` used to mean `-` (schedule early) and `warn` (the command to
change log level)`.

`sbt-extras` translates its `-w` option into `--warn`, or passes through
`--warn` if provided. The official SBT launcher passes through `--warn`.

`sbt -warn` still works in 1.x, but `sbt --warn` doesn't give the non-zero
error code after a failure.

```
for sbt in 0.13.17 1.1.4; do for opt in "--warn" "-warn"; do (set -x; java -jar /home/jenkins/.sbt/launchers/0.13.17/sbt-launch.jar -Dsbt.version=$sbt $opt 'eval ???'; echo $?;) done; done
+ java -jar /home/jenkins/.sbt/launchers/0.13.17/sbt-launch.jar -Dsbt.version=0.13.17 --warn 'eval ???'
scala.NotImplementedError: an implementation is missing
    at scala.Predef$.$qmark$qmark$qmark(Predef.scala:252)
...
[error] scala.NotImplementedError: an implementation is missing
[error] Use 'last' for the full log.
+ echo 1
1
+ java -jar /home/jenkins/.sbt/launchers/0.13.17/sbt-launch.jar -Dsbt.version=0.13.17 -warn 'eval ???'
scala.NotImplementedError: an implementation is missing
    at scala.Predef$.$qmark$qmark$qmark(Predef.scala:252)
...
[error] scala.NotImplementedError: an implementation is missing
[error] Use 'last' for the full log.
+ echo 1
1
+ java -jar /home/jenkins/.sbt/launchers/0.13.17/sbt-launch.jar -Dsbt.version=1.1.4 --warn 'eval ???'
[warn] sbt version mismatch, current: 1.1.4, in build.properties: "0.13.17", use 'reboot' to use the new value.
[info] Loading project definition from /tmp/scratch/project
[info] Updating ProjectRef(uri("file:/tmp/scratch/project/"), "scratch-build")...
[info] Done updating.
[info] Set current project to scratch (in build file:/tmp/scratch/)
[warn] The `-` command is deprecated in favor of `onFailure` and will be removed in a later version
[error] scala.NotImplementedError: an implementation is missing
[error]     at scala.Predef$.$qmark$qmark$qmark(Predef.scala:284)
..
[error] scala.NotImplementedError: an implementation is missing
[error] Use 'last' for the full log.
+ echo 0
0
+ java -jar /home/jenkins/.sbt/launchers/0.13.17/sbt-launch.jar -Dsbt.version=1.1.4 -warn 'eval ???'
+ echo 1
1
```

This commits changes our scripts to use the recommended `-warn`. However,
I'd say that SBT 1.x should either reject `--warn` outright or be compatible
with the old behaviour.

adriaanm commented:
Rebased once more, with date ordering for github and more squashiness.

SethTisue commented:
discussed at team meeting today — nobody had objections to merging this as soon as CI is green.

adriaanm commented:
/synch

adriaanm commented:
/rebuild

eed3si9n commented:
Looks green to me.

SethTisue commented:
ahhhhhhhhhh that makes my day.

jvican commented on test/osgi/src/ScalaOsgiHelper.scala:
```diff
@@ -7,6 +7,7 @@ import java.io.File
 trait ScalaOsgiHelper {
 
   private def allBundleFiles = {
+    println(new File(".").getAbsolutePath)
```
You may want to remove this :smile:

SethTisue commented on test/benchmarks/project/build.properties:
```diff
@@ -0,0 +1 @@
+sbt.version=1.1.4
```
nbd, but we should bump this one to 1.2.3 too

xuwei-k commented:
https://github.com/scala/scala/pull/7274
